cpp/src/randomforest/randomforest.cu

```diff
@@ -241,79 +359,99 @@ rfClassifier<T>::rfClassifier(RF_params cfg_rf_params)
+    handle.getDeviceAllocator(), stream, temp_storage_bytes);
+  std::shared_ptr<TemporaryMemory<T, int>> tempmem =
+    std::make_shared<TemporaryMemory<T, int>>(
+      user_handle.getImpl(), n_sampled_rows, n_cols, 1, n_unique_labels,
+      this->rf_params.tree_params.n_bins,
+      this->rf_params.tree_params.split_algo);
+  if ((this->rf_params.tree_params.split_algo == SPLIT_ALGO::GLOBAL_QUANTILE) && !(this->rf_params.tree_params.quantile_per_tree)) {
+    preprocess_quantile(input, nullptr, n_sampled_rows, n_cols, n_rows,
+                        this->rf_params.tree_params.n_bins, tempmem);
+  }
   for (int i = 0; i < this->rf_params.n_trees; i++) {
-    // Select n_sampled_rows (with replacement) numbers from [0, n_rows) per tree.
-    // selected_rows: randomly generated IDs for bootstrapped samples (w/ replacement); a device ptr.
-    MLCommon::device_buffer<unsigned int> selected_rows(
-      handle.getDeviceAllocator(), stream, n_sampled_rows);
-
-    if (this->rf_params.bootstrap) {
-      MLCommon::Random::Rng r(
-        i *
-        1000);  // Ensure the seed for each tree is different and meaningful.
-      r.uniformInt(selected_rows.data(), n_sampled_rows, (unsigned int)0,
-                   (unsigned int)n_rows, stream);
-    } else {
-      std::vector<unsigned int> h_selected_rows(n_rows);
-      std::iota(h_selected_rows.begin(), h_selected_rows.end(), 0);
-      std::random_shuffle(h_selected_rows.begin(), h_selected_rows.end());
-      h_selected_rows.resize(n_sampled_rows);
-      MLCommon::updateDevice(selected_rows.data(), h_selected_rows.data(),
-                             n_sampled_rows, stream);
-    }
+    this->prepare_fit_per_tree(handle, i, n_rows, n_sampled_rows,
+                               selected_rows.data(),
+                               sorted_selected_rows.data(),
+                               rows_temp_storage->data(), temp_storage_bytes);
 
     /* Build individual tree in the forest.
 		   - input is a pointer to orig data that have n_cols features and n_rows rows.
 		   - n_sampled_rows: # rows sampled for tree's bootstrap sample.
-		   - selected_rows: points to a list of row #s (w/ n_sampled_rows elements) used to build the bootstrapped sample.
-			Expectation: Each tree node will contain (a) # n_sampled_rows and (b) a pointer to a list of row numbers w.r.t original data.
+		   - sorted_selected_rows: points to a list of row #s (w/ n_sampled_rows elements) used to build the bootstrapped sample.
+		   Expectation: Each tree node will contain (a) # n_sampled_rows and (b) a pointer to a list of row numbers w.r.t original data.
 		*/
-    this->trees[i].fit(user_handle, input, n_cols, n_rows, labels,
-                       selected_rows.data(), n_sampled_rows, n_unique_labels,
-                       this->rf_params.tree_params);
 
-    //Cleanup
-    selected_rows.release(stream);
+    trees[i].fit(user_handle, input, n_cols, n_rows, labels,
```

Comment:
I think decision tree building should be parallelized to get better speedup. Should be relatively straigforward using OpenMP with multiple streams. It can be done in a separate PR.

Reply:
Agreed but as per the cuml developer guide. We want to keep the algorithms single threaded. Openmp can be compiler specific.

Reply:
@vishalmehta1991, Some of us had a discussion about this on #694. When I get some time, I'm going to submit a PR so that we can be more clear in our language in the developer guide.

This is a case that would warrant the use of threading since building trees in serial has such a massive impact on the result of our application. This should be made possible by synchronizing the main stream before your for..loop and creating new streams in threads and synchronizing them while still in the for loop. I have also verified with the CUDA team that the cuda API is, in fact, thread safe so you should not have a problem creating and submitting cuda kernels while in separate threads.

Reply:
Hi @cjnolet am completely aware about the CUDA api and the thread support for it.  That not the concern. My concern is that the way openmp is treated across a board of compilers. To be compliment with all the compiler that do openmp; we will need to test them. 

But again from a developer point; putting openmp is easy and simple. I would be up for it if Rapids developer guide allows me to.

Reply:
Tagging @jirikraus for additional feedback.

Reply:
@Vishalmehta1991. I'm having trouble finding documentation on the compiler problems. Would you mind sharing some references with me?

I'm concerned that we're seeing signs the current implementation's performance is bounded by the number of available CPU cores as a result of the kernels being submitted in a single stream.

Reply:
Seriously, what is the real issue here about OpenMP or any other multi-threading library? OpenMP was used in KNN and I don't recall any problem. It's obvious that we need to build the trees in parallel. What do you guys suggest if you don't want to use openmp or a multi-threading library? Only other option I can think of is to redesign and rewrite the whole algorithm from scratch to build the trees in parallel which will require a significant amount of work. Any other suggestions?

Reply:
Personally I dont mind threads; its embarrassing parallel. If you want we can do openmp in a separate PR (this PR is too big ) and also open a PR to change the cuml developer guide; so that we adhere to cuml guidelines.

Reply:
Let's merge this PR and try openmp in a separate PR. We can change the developer guideline if openmp PR gives us good results.

Reply:
The wording that is being discussed for our changes to the developer guide: 
```
With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, assumed to be single threaded. Exceptions are made for algorithms that can take advantage of multiple CUDA streams in order to oversubscribe or increase occupancy on a single GPU. In these cases, multiple threads should be used only to maintain concurrency of the underlying CUDA streams. Multiple threads should be used sparingly, be bounded, and should not perform CPU-intensive computations.
```